File: drivers/lua/ssd1306/ssd1306.lua
-- ssd1306.lua -- SSD1306 OLED driver (host). Mirror of C `ssd1306.[ch]`:
-- identical draw logic; only the transport (i2c_host Bus) and the show()
-- chunking differ (see show()).
--
-- Mono OLED on I2C (XIAO expansion board: 128x64 @ 0x3C). Every transfer is a
-- control byte then a stream: 0x00 = command, 0x40 = data. RAM = W cols x
-- (H/8) pages, one byte = 8 vertical pixels (bit0 = top). Draw into a local
-- framebuffer, then show() the whole thing.
--
--   luajit ssd1306.lua [--port /dev/ttyACMx] [--kind pico|samd21] [--addr 0x3C]

local _dir = (debug.getinfo(1, "S").source:sub(2)):match("(.*/)") or "./"
local _root = _dir .. "../"                        -- drivers/lua/ (each driver is one level under)
package.path = _root .. "i2c/?.lua;" .. _root .. "pcf8563/?.lua;" .. _root .. "ssd1306/?.lua;"
    .. _root .. "../../tools/commission/lua/?.lua;" .. package.path
local i2c  = require("i2c_host")
local font = require("font5x7")
local bit  = require("bit")

local CTRL_CMD, CTRL_DATA = 0x00, 0x40

local SSD = {}
SSD.__index = SSD
local M = {}

function M.new(bus, addr, w, h)
    local self = setmetatable({ bus = bus, addr = addr or 0x3C, w = w or 128, h = h or 64 }, SSD)
    self.pages = self.h / 8
    self.fb = {}
    self:clear()
    return self
end

-- command stream (0x00 control byte)
function SSD:_cmd(bytes) self.bus:write(self.addr, string.char(CTRL_CMD) .. bytes) end

function SSD:init()
    local mux     = self.h - 1
    local compins = (self.h == 32) and 0x02 or 0x12
    self:_cmd(string.char(
        0xAE,             -- display off
        0xD5, 0x80,       -- clock divide
        0xA8, mux,        -- multiplex = H-1
        0xD3, 0x00,       -- display offset 0
        0x40,             -- start line 0
        0x8D, 0x14,       -- charge pump on
        0x20, 0x00,       -- horizontal addressing mode
        0xA1,             -- segment remap
        0xC8,             -- COM scan reversed -> upright
        0xDA, compins,    -- COM pins config
        0x81, 0xCF,       -- contrast
        0xD9, 0xF1,       -- pre-charge
        0xDB, 0x40,       -- VCOMH deselect
        0xA4,             -- resume to RAM
        0xA6,             -- normal (not inverted)
        0xAF))            -- display on
    self:clear()
    self:show()
end

function SSD:clear()
    for i = 1, self.w * self.pages do self.fb[i] = 0 end
end

function SSD:pixel(x, y, on)
    if x < 0 or y < 0 or x >= self.w or y >= self.h then return end
    local idx  = math.floor(y / 8) * self.w + x + 1   -- 1-based framebuffer index
    local mask = bit.lshift(1, y % 8)
    if on then self.fb[idx] = bit.bor(self.fb[idx], mask)
    else       self.fb[idx] = bit.band(self.fb[idx], bit.band(bit.bnot(mask), 0xFF)) end
end

function SSD:char(x, y, ch)
    local c = ch:byte()
    if c < 0x20 or c > 0x7E then c = 0x20 end
    local g = font[c - 0x20 + 1]
    for col = 0, 4 do
        local bits = g[col + 1]
        for row = 0, 6 do
            if bit.band(bits, bit.lshift(1, row)) ~= 0 then self:pixel(x + col, y + row, true) end
        end
    end
end

function SSD:text(x, y, s)
    for i = 1, #s do self:char(x, y, s:sub(i, i)); x = x + 6 end
end

function SSD:show()
    self:_cmd(string.char(0x21, 0, self.w - 1, 0x22, 0, self.pages - 1))  -- col + page range
    -- The framebuffer ships as data (0x40) chunked to the transport's write
    -- cap (the 32/64 B USB-shell limit). Each chunk = 0x40 + up to max_write-1
    -- framebuffer bytes. (The C twin sends it all in one chip-native write.)
    local n     = self.w * self.pages
    local chunk = self.bus.max_write - 1
    local i = 1
    while i <= n do
        local last, parts = math.min(i + chunk - 1, n), {}
        for k = i, last do parts[#parts + 1] = string.char(self.fb[k]) end
        self.bus:write(self.addr, string.char(CTRL_DATA) .. table.concat(parts))
        i = last + 1
    end
end

function SSD:display_on(on) self:_cmd(string.char(on and 0xAF or 0xAE)) end
function SSD:invert(inv)    self:_cmd(string.char(inv and 0xA7 or 0xA6)) end
function SSD:contrast(c)    self:_cmd(string.char(0x81, c)) end

M.SSD = SSD

-- ---- CLI: visual self-test --------------------------------------------------
if arg and (arg[0] or ""):match("ssd1306%.lua$") then
    local port, kind, addr
    local i = 1
    while arg[i] do
        if arg[i] == "--port" then i = i + 1; port = arg[i]
        elseif arg[i] == "--kind" then i = i + 1; kind = arg[i]
        elseif arg[i] == "--addr" then i = i + 1; addr = tonumber(arg[i]) end
        i = i + 1
    end
    local function sleep(s) local t = os.time() + s; repeat until os.time() >= t end

    local bus = i2c.open(port, kind)
    print(string.format("%s front-end on %s", kind or "pico", bus.port))
    local oled = M.new(bus, addr or 0x3C)
    print(string.format("SSD1306 @ 0x%02X (%dx%d) -- visual self-test", oled.addr, oled.w, oled.h))

    oled:init()
    oled:clear()
    for x = 0, oled.w - 1 do oled:pixel(x, 0, true); oled:pixel(x, oled.h - 1, true) end   -- border
    for y = 0, oled.h - 1 do oled:pixel(0, y, true); oled:pixel(oled.w - 1, y, true) end
    oled:text(4, 4,  "slow_bus SSD1306")
    oled:text(4, 14, "0123456789 +-*/=")
    oled:text(4, 24, "ABCDEFGHIJKLMNOPQR")
    oled:text(4, 34, "abcdefghijklmnopqr")
    oled:text(4, 44, "#$%&@?!<>[]{}~^")
    oled:show()
    print("  drew border + 5 text rows")

    sleep(1); oled:invert(true);  print("  invert ON")
    sleep(1); oled:invert(false); print("  invert OFF")

    print("\n  look at the OLED: a full border with 5 rows of text, that flashed inverted once.")
    bus:close()
end

return M

File: drivers/c/ssd1306/ssd1306.c
// SSD1306 OLED driver (C). Mirror of `ssd1306.lua` -- identical draw logic;
// only the transport (i2c_bus_t) and the show() chunking differ. See the note
// on show() for why this side does NOT chunk.
#include "ssd1306.h"
#include "font5x7.h"
#include <string.h>

#define CTRL_CMD   0x00u
#define CTRL_DATA  0x40u

// command stream: [0x00][cmd bytes...]. The init list is short, so one write.
static bool cmd(ssd1306_t *d, const uint8_t *c, size_t n) {
    uint8_t b[40];
    if (n + 1 > sizeof b) return false;
    b[0] = CTRL_CMD;
    memcpy(b + 1, c, n);
    return i2c_bus_write(d->bus, d->addr, b, n + 1) == 0;
}

bool ssd1306_init(ssd1306_t *d, const i2c_bus_t *bus, uint8_t addr, uint8_t w, uint8_t h) {
    d->bus  = bus;
    d->addr = addr ? addr : SSD1306_ADDR;
    d->w    = w ? w : 128u;
    d->h    = h ? h : 64u;
    d->pages = (uint8_t)(d->h / 8u);
    const uint8_t mux     = (uint8_t)(d->h - 1u);
    const uint8_t compins = (d->h == 32u) ? 0x02u : 0x12u;
    const uint8_t init[] = {
        0xAE,                 // display off
        0xD5, 0x80,           // clock divide
        0xA8, mux,            // multiplex = H-1
        0xD3, 0x00,           // display offset 0
        0x40,                 // start line 0
        0x8D, 0x14,           // charge pump on
        0x20, 0x00,           // horizontal addressing mode
        0xA1,                 // segment remap (column 127 -> SEG0)
        0xC8,                 // COM scan direction reversed -> upright
        0xDA, compins,        // COM pins config
        0x81, 0xCF,           // contrast
        0xD9, 0xF1,           // pre-charge
        0xDB, 0x40,           // VCOMH deselect
        0xA4,                 // resume to RAM content
        0xA6,                 // normal (not inverted)
        0xAF,                 // display on
    };
    if (!cmd(d, init, sizeof init)) return false;
    ssd1306_clear(d);
    return ssd1306_show(d);
}

void ssd1306_clear(ssd1306_t *d) {
    memset(SSD1306_FB(d), 0, (size_t)d->w * d->pages);
}

void ssd1306_pixel(ssd1306_t *d, int x, int y, bool on) {
    if (x < 0 || y < 0 || x >= d->w || y >= d->h) return;
    uint8_t *fb = SSD1306_FB(d);
    unsigned idx = (unsigned)(y >> 3) * d->w + (unsigned)x;
    uint8_t bit = (uint8_t)(1u << (y & 7));
    if (on) fb[idx] |= bit;
    else    fb[idx] = (uint8_t)(fb[idx] & ~bit);
}

void ssd1306_char(ssd1306_t *d, int x, int y, char c) {
    unsigned ch = (unsigned char)c;
    if (ch < 0x20u || ch > 0x7Eu) ch = 0x20u;
    const uint8_t *g = font5x7[ch - 0x20u];
    for (int col = 0; col < 5; col++)
        for (int row = 0; row < 7; row++)
            if (g[col] & (1u << row)) ssd1306_pixel(d, x + col, y + row, true);
}

void ssd1306_text(ssd1306_t *d, int x, int y, const char *s) {
    for (; *s; s++, x += 6) ssd1306_char(d, x, y, *s);
}

bool ssd1306_show(ssd1306_t *d) {
    const uint8_t addr_cmd[] = { 0x21, 0x00, (uint8_t)(d->w - 1u),       // column range
                                 0x22, 0x00, (uint8_t)(d->pages - 1u) }; // page range
    if (!cmd(d, addr_cmd, sizeof addr_cmd)) return false;
    // The framebuffer ships in ONE write: buf[0] is already the 0x40 data
    // control byte. A chip-native I2C master has no transfer cap (the 32/64 B
    // limit is a host USB-shell artifact, handled on the Lua side), so unlike
    // ssd1306.lua this does not chunk.
    d->buf[0] = CTRL_DATA;
    return i2c_bus_write(d->bus, d->addr, d->buf, 1u + (size_t)d->w * d->pages) == 0;
}

bool ssd1306_display_on(ssd1306_t *d, bool on) {
    uint8_t c = on ? 0xAFu : 0xAEu;
    return cmd(d, &c, 1);
}
bool ssd1306_invert(ssd1306_t *d, bool inv) {
    uint8_t c = inv ? 0xA7u : 0xA6u;
    return cmd(d, &c, 1);
}
bool ssd1306_contrast(ssd1306_t *d, uint8_t c) {
    uint8_t b[2] = { 0x81u, c };
    return cmd(d, b, sizeof b);
}
